cluster-autoscaler/simulator/delta_cluster_snapshot.go

```diff
@@ -0,0 +1,484 @@
+		data.deletedNodeInfos[nodeName] = true
+	}
+
+	if !foundInBase && !foundInDelta {
+		// Node not found in the chain.
+		return errNodeNotFound
+	}
+
+	// Maybe consider deleting from the lists instead. Maybe not.
+	data.clearCaches()
+	return nil
+}
+
+func (data *internalDeltaSnapshotData) addPod(pod *apiv1.Pod, nodeName string) error {
+	dni, found := data.getNodeInfoLocal(nodeName)
+	if !found {
+		bni, found := data.baseData.getNodeInfo(nodeName)
+		if !found {
+			return errNodeNotFound
+		}
+		dni = bni.Clone()
+		data.modifiedNodeInfoMap[nodeName] = dni
+		data.clearCaches()
+	}
+
+	dni.AddPod(pod)
+	if data.podList != nil || data.havePodsWithAffinity != nil {
+		data.clearPodCaches()
+	}
+	return nil
+}
+
+func (data *internalDeltaSnapshotData) removePod(namespace, name, nodeName string) error {
+	dni, found := data.getNodeInfoLocal(nodeName)
+	if !found {
+		bni, found := data.baseData.getNodeInfo(nodeName)
+		if !found {
+			return errNodeNotFound
+		}
+		dni = bni.Clone()
+		data.modifiedNodeInfoMap[nodeName] = dni
+		data.clearCaches()
+	}
+
+	preAffinityPods := len(dni.PodsWithAffinity())
+
+	podFound := false
+	for _, pod := range dni.Pods() {
+		if pod.Namespace == namespace && pod.Name == name {
+			if err := dni.RemovePod(pod); err != nil {
+				return fmt.Errorf("cannot remove pod; %v", err)
+			}
+			podFound = true
+			break
+		}
+	}
+	if !podFound {
+		return fmt.Errorf("pod %s/%s not in snapshot", namespace, name)
+	}
+
+	// Maybe consider deleting from the list in the future. Maybe not.
+	postAffinityPods := len(dni.PodsWithAffinity())
+	if preAffinityPods == 1 && postAffinityPods == 0 {
+		data.havePodsWithAffinity = nil
+	}
+	if data.podList != nil {
+		data.clearPodCaches()
+	}
+	return nil
+}
+
+func (data *internalDeltaSnapshotData) getPodList() []*apiv1.Pod {
+	if data == nil {
+		return nil
+	}
+	if data.podList == nil {
+		data.podList = data.buildPodList()
+	}
+	return data.podList
+}
+
+func (data *internalDeltaSnapshotData) buildPodList() []*apiv1.Pod {
+	podList := []*apiv1.Pod{}
+	if len(data.deletedNodeInfos) > 0 || len(data.modifiedNodeInfoMap) > 0 {
+		nodeInfos := data.getNodeInfoList()
+		for _, ni := range nodeInfos {
+			podList = append(podList, ni.Pods()...)
+		}
+	} else {
+		basePodList := data.baseData.getPodList()
+		copy(podList, basePodList)
+		for _, ni := range data.addedNodeInfoMap {
+			podList = append(podList, ni.Pods()...)
+		}
+	}
+	return podList
+}
+
+func (data *internalDeltaSnapshotData) fork() *internalDeltaSnapshotData {
+	forkedData := newInternalDeltaSnapshotData()
+	forkedData.baseData = data
+	return forkedData
+}
+
+func (data *internalDeltaSnapshotData) commit() *internalDeltaSnapshotData {
+	for _, node := range data.modifiedNodeInfoMap {
+		data.baseData.updateNode(node)
+	}
+	for _, node := range data.addedNodeInfoMap {
+		data.baseData.addNodeInfo(node)
+	}
+	for node := range data.deletedNodeInfos {
+		data.baseData.removeNode(node)
+	}
```

Comment:
Invalidate caches on baseData

Reply:
I'm having second thoughts about it :s If we're using the same methods as for other updates, we should be able to trust those methods to invalidate caches if needed.

Reply:
And yet I don't :) As per my earlier comment addNodeInfo must either drop pod caches or explicitly assume nodeInfo it adds have no pods (in which case you should add a comment stating it and drop pod caches here).

Problematic sequence (maybe add some list calls to tests you already have to make them verify cache invalidation?):

snapshot.Pods().List()
snapshot.Fork()
snaphost.AddNode(node)
snapshot.AddPod(pod, node.Name())
snapshot.Commit()
snapshot.Pods().List() # my theory is that this will miss pod we added in fork